src/zenml/utils/function_utils.py

```diff
@@ -0,0 +1,221 @@
+#  Copyright (c) ZenML GmbH 2024. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Utility functions for python functions."""
+
+import inspect
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Any, Callable, Iterator, List, Tuple, TypeVar, Union
+
+import click
+
+from zenml.logger import get_logger
+from zenml.utils.string_utils import random_str
+
+F = TypeVar("F", bound=Callable[..., None])
+
+logger = get_logger(__name__)
+
+_CLI_WRAPPED_SCRIPT_TEMPLATE_HEADER = """
+from zenml.utils.function_utils import _cli_wrapped_function
+
+import sys
+sys.path.append("{func_path}")
+from {func_module} import {func_name} as func_to_wrap
+
+if entrypoint:=getattr(func_to_wrap, "entrypoint", None):
+    func = _cli_wrapped_function(entrypoint)
+else:
+    func = _cli_wrapped_function(func_to_wrap)
+
+"""
+_CLI_WRAPPED_ACCELERATE_MAIN = """
+if __name__=="__main__":
+    from accelerate import Accelerator
+    import cloudpickle as pickle
+
+    accelerator = Accelerator()
+
+    ret = func(standalone_mode=False)
+
+    if accelerator.is_main_process:
+        pickle.dump(ret, open("{output_file}", "wb"))
+"""
+_ALLOWED_TYPES = (str, int, float, bool, Path)
+_ALLOWED_COLLECTIONS = (tuple,)
+_TYPES_MAPPER = {
+    str: click.STRING,
+    int: click.INT,
+    float: click.FLOAT,
+    bool: click.BOOL,
+    Path: click.STRING,
+    None: click.STRING,
+}
+
+
+def _cli_arg_name(arg_name: str) -> str:
+    return arg_name.replace("_", "-")
+
+
+def _is_valid_collection_arg(arg_type: Any) -> bool:
+    if getattr(arg_type, "__origin__", None) in _ALLOWED_COLLECTIONS:
+        if arg_type.__args__[0] not in _ALLOWED_TYPES:
+            return False
+        return True
+    return False
+
+
+def _is_valid_optional_arg(arg_type: Any) -> bool:
```

Comment:
```suggestion
def _is_valid_optional_arg(arg_type: Any) -> bool:
    """Check if the given argument type is a valid Optional type.

    A valid Optional type is defined as a Union with two arguments, where:
    - The first argument is either an allowed type or a valid collection type.
    - The second argument is the NoneType.

    Args:
        arg_type: The type to check.

    Returns:
        True if the argument type is a valid Optional type, False otherwise.
    """
```